index.js

```diff
@@ -28,10 +28,36 @@ function apply(options, compiler) {
 
   var runner = runCompilation.bind(this, options);
 
-  compiler.plugin('run', runner);
-  compiler.plugin('watch-run', function onWatchRun(watcher, callback) {
-    runner(watcher.compiler, callback);
-  });
+  if (options.lintDirtyModulesOnly) {
+    this.startTime = Date.now();
+    this.prevTimestamps = {};
+    var isFirstRun = true;
+
+    compiler.plugin('emit', function (compilation, callback) {
+      var dirtyOptions = assign({}, options);
+      var globPatterb = dirtyOptions.files;
+
+      var changedFiles = Object.keys(compilation.fileTimestamps).filter(function (watchfile) {
+        return (this.prevTimestamps[watchfile] || this.startTime) < (compilation.fileTimestamps[watchfile] || Infinity);
+      }.bind(this)).filter(minimatch.filter(globPatterb.join('|'), {matchBase: true}));
```

Comment:
Sorry, did not quite get where to put reduce. Would this be some like this?
```js
function getChangedFiles(plugin, compilation, glob) {
    var updatedStyles = reduce(compilation.fileTimestamps, function (changedStyleFiles, timestamp, filename) {
      if (
        // Check if file has been changed first ...
        (plugin.prevTimestamps[filename] || plugin.startTime) < (compilation.fileTimestamps[filename] || Infinity) &&
        // ... then validate by the glob pattern.
        minimatch(glob, {matchBase: true})
      ) {
        changedStyleFiles.push(filename);
      }
      return changedStyleFiles;
    }, []);
      
    return updatedStyles;
  }
```

An example without reduce:
```js
function getChangedFiles(plugin, compilation, glob) {
    // Filter only changed files first ...
    var updated = Object.keys(compilation.fileTimestamps).filter(function (filename) {
      return (plugin.prevTimestamps[filename] || plugin.startTime) < (compilation.fileTimestamps[filename] || Infinity);
    });
    // ... then filter files by the glob pattern.
    var updatedStyles = updated.filter(minimatch.filter(glob, {matchBase: true}));
    return updatedStyles;
  }
```

The actual call would look like:
```js
function apply(options, compiler) {
  var plugin = this;
  ...
  compiler.plugin('emit', function (compilation, callback) {
     ...
     var changedFiles = getChangedFiles(plugin, compilation, globPatterb.join('|'))
     ...
  }
  ...
}
```

@JaKXz, Let me what option you'd prefer to go with